instance/nomad_client.py

```diff
@@ -0,0 +1,91 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Convenience wrappers for calls to the Nomad REST API.
+"""
+
+# Imports #####################################################################
+
+import functools
+
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+import requests
+
+
+# Functions ###################################################################
+
+def request(method, *endpoint_parts, server=settings.NOMAD_SERVER, version="v1", **kwargs):
+    """Send a REST API request to the Nomad API and return the Response object."""
+    if server is None:
+        raise ImproperlyConfigured("Please set NOMAD_SERVER in the Django settings.")
+    url = "https://" + "/".join([server, version, *endpoint_parts])
+    kwargs.setdefault("verify", settings.NOMAD_CACERT)
+    kwargs.setdefault("cert", (settings.NOMAD_CLIENT_CERT, settings.NOMAD_CLIENT_KEY))
+    kwargs.setdefault("timeout", 15)
+    return requests.request(method, url, **kwargs)
+
+# Convenience wrappers for specific HTTP methods
+get = functools.partial(request, "get")
+post = functools.partial(request, "post")
+put = functools.partial(request, "put")
+delete = functools.partial(request, "delete")
+
+
+# Classes #####################################################################
+
+class NomadJob:
+    """Class to represent a Nomad job.
+
+    This representation is meant to be ephemeral, and it is not persisted to the database.
+    """
+
+    def __init__(self, jobspec):
+        """Initialize a NomadJob instance.
+
+        The jobspec parameter is expected to be a dictionary that can be serialized to a valid JSON
+        jobspec.
+        """
+        self.jobspec = jobspec
+        self.job_id = jobspec["ID"]
+
+    def run(self):
+        """Create the job on the Nomad cluster."""
+        return post("jobs", json={"Job": self.jobspec})
+
+    def stop(self):
+        """Delete the job from the Nomad cluster."""
+        return delete("job", self.job_id)
+
+    def all_running(self):
+        """Return a Boolean value indicating whether all desired task groups are running."""
+        # This function is somewhat improvised.  The status of a Nomad job is rather complex, since
+        # all task groups and their allocations have their own status.  This implementation aims
+        # to be a reasonable simple way of determining whether the job started successfully.  I
+        # would prefer to rely on Consul health checks instead, but currently I don't know a way of
+        # making them work properly together with Consul Connect.
```

Comment:
I think the current technique is sufficient for now. I don't recall having any behavioral issues with Redis/Memcache themselves if their processes are already successfully running - they're quite reliable systems.

Reply:
The only concern would be Nomad disposing of jobs because they hit resource limits, but I don't think this actually happens for the `raw_exec` driver.

Reply:
In that case maybe the OOM killer would step in instead and bring down some services. In general I'm a bit worried now that if Nomad doesn't check resource limits, we could have a single customer somehow storing too much into their Memcached and thus causing issues for other services on the same machine. I guess we can't get everything right but something to think about.